Workflow fixes related to DIP Upload & Store

jrwdunham opened the pull request:
Changes to the workflow and related client scripts so that the "Upload DIP" and "Store DIP" decision points make more sense and allow for preconfigured choices for "Store DIP" AFTER "Upload DIP". See RM issue 11195 and closed PR #672.
    
* There used to be 2 distinct "Store DIP" paths—now there is only 1.
* handleUnstoredDIP client script puts an unstored DIP in the rejected/ directory *only* if it has not been copied to the uploadedDIPs/ directory
* Processing config form allows configuration of "Store DIP" decision
* Microsecond precision on SIP status timestamp: Now when ingest/status is called by jobs.js, the dashboard returns microsecond precision on the unix timestamps for SIP last modified. This fixes a bug where the representation of a SIP will not update in the GUI if its final jobs all completed at the same second.

jrwdunham committed:
Microsecond precision on SIP status timestamp

Now when ingest/status is called by jobs.js, the dashboard returns
microsecond precision on the unix timestamps for SIP last modified. This
fixes a bug where the representation of a SIP will not update in the GUI if its
final jobs all completed at the same second.

sevein commented on src/dashboard/src/components/ingest/views.py:
```diff
@@ -90,61 +90,70 @@ def post(self, request):
+    """
+    sql = """
+        SELECT SIPUUID,
+                MAX(UNIX_TIMESTAMP(createdTime) + createdTimeDec) AS timestamp
```
Nice :)

sevein commented on src/MCPClient/lib/clientScripts/handleUnstoredDIP.py:
```diff
@@ -0,0 +1,76 @@
+
+def _sip_uploaded(sip_path, uploaded_dips_path):
+    """Return ``True`` if the SIP has been uploaded."""
+    return _get_sip_dirname(sip_path) in os.listdir(uploaded_dips_path)
```
Is there any reason why you can't just use `os.path.isdir(path)` or `os.path.exists(path)`? Sounds like it'd be more efficient because it just uses `stat()`.

jrwdunham replied:
You're right—I'll use `isdir`.

sevein commented on src/dashboard/src/main/migrations/0038_fix_workflow_store_dip.py:
```diff
@@ -0,0 +1,189 @@
+
+    Steps:
+
+    1. Remove "Reject DIP" from options at "Upload DIP" decision point:
```
`s/:/.`?

jrwdunham replied:
ok. fixing punctuation

sevein commented on src/dashboard/src/main/migrations/0038_fix_workflow_store_dip.py:
```diff
@@ -0,0 +1,189 @@
+         (e3efab02-1860-42dd-a46c-25601251b930)
+
+    4. Change "Completed" link to "Handle unstored DIP". This is handled by a
+       new client script that moves the DIP to the rejected/ directory iff it
```
`s/iff/if`?

sevein replied:
Joel explained to me that this is the way that biconditionals are abbreviated in English - makes sense! :)

jrwdunham replied:
intentional

sevein commented on src/dashboard/src/main/migrations/0038_fix_workflow_store_dip.py:
```diff
@@ -0,0 +1,189 @@
+    # deletion of the three that follow, despite the fact that this does not
+    # seem to be entailed by Django's foreign key cascade delete strategy.
+    StandardTaskConfig.objects.get(id=reject_dip_stc_uuid).delete()
+    TaskConfig.objects.get(id=reject_dip_tc_uuid).delete()
```
`TaskConfig` is referenced by `MicroServiceChainLink. currenttask`. `MicroServiceChainLink` is referenced by `MicroServiceChainChoice. choiceavailableatlink` and so on. So I'd assume that the behaviour you're seeing is because of Django is emulating `ON DELETE CASCADE`?

jrwdunham replied:
Ok. I modified the comment to reference the Django docs. I'm still leaving the other delete attempts in there in case deleting the first two does not delete all of the last three.

sevein commented on src/dashboard/src/main/migrations/0038_fix_workflow_store_dip.py:
```diff
@@ -0,0 +1,189 @@
+    completed_tc.save()
+    completed_stc = StandardTaskConfig.objects.get(id=completed_stc_uuid)
+    completed_stc.arguments = ('"%SIPDirectory%" '
+                               '"%rejectedDirectory%." '
```
Is the dot necessary?

jrwdunham replied:
No, no I don't believe it is.

sevein commented on src/dashboard/src/main/migrations/0038_fix_workflow_store_dip.py:
```diff
@@ -0,0 +1,189 @@
+class Migration(migrations.Migration):
+
+    dependencies = [('main', '0037_mediaconch_policy_checks')]
+    operations = [migrations.RunPython(data_migration)]
```
This is a clear style deviation from Django's suggested way, not that I care but I find interesting that you prefer your way over the convention :D

jrwdunham replied:
I didn't know about Django's style guide on this issue. I'll conform to it. Just thought the whitespace wasn't necessary in this case.

sevein reviewed (changes requested):
Please see my comments above.

jrwdunham committed:
Workflow fixes re: Upload/Store DIP

Changes to the workflow and related client scripts so that the "Upload
DIP" and "Store DIP" decision points make more sense and allow for
preconfigured choices for "Store DIP" AFTER "Upload DIP":

- There used to be 2 distinct "Store DIP" paths---now there is only 1.
- handleUnstoredDIP client script puts an unstored DIP in the rejected/
  directory *only* if it has not been copied to the uploadedDIPs/
  directory
- Processing config form allows configuration of "Store DIP" decision
